File: javascript/src/iterator/value/AbstractIteratorValue.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {IteratorValue} from "./IteratorValue"

/**
 * A generalization of a {@link IteratorValue} to have a common ancestor.
 *
 * @typeParam T       The element type
 * @typeParam IS_DONE Whenever the {@link IteratorValue} has finished or not (`boolean` by default)
 */
export abstract class AbstractIteratorValue<const T,
    const IS_DONE extends boolean = boolean, >
    implements IteratorValue<T> {

    protected constructor() {}


    public abstract isDone(): IS_DONE

    public get done(): IS_DONE { return this.isDone() }

    public isNotDone(): boolean { return !this.isDone() }

    public abstract get value(): T

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/iterator/value/AbstractIteratorValue.java
package joookiwi.collection.java.iterator.value;

import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;

///A generalization of a [IteratorValue] to have a common ancestor.
///
/// @param <T> The element type
@NotNullByDefault
public abstract class AbstractIteratorValue<T extends @Nullable Object>
        implements IteratorValue<T> {

    protected AbstractIteratorValue() { super(); }


    @Override public final boolean done() { return isDone(); }

    @Override public boolean isNotDone() { return !isDone(); }

}
